### PR TITLE
Add LFlip method 

### PR DESCRIPTION
**Pull request by arthurPignet**
Add the LFlip method
The method is the previously denoted EM - method.

### CODE REVIEW

File: mplc/multi_partner_learning.py [hunk 1]
```diff
@@ -654,3 +655,122 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
...
+                    self.theta_[self.epoch_index] = self.theta[self.epoch_index]  # Initialize the theta_
+                    for y in range(self.K):  # TODO ? vectorize
+                        for idx, x in enumerate(x_batch):
+                            self.theta_[self.epoch_index,
+                                        partner_index,
+                                        y,
+                                        np.argmax(y_batch[idx])] *= np.max(partner_model.predict(np.expand_dims(x,
+                                                                                                                axis=0)
+                                                                                                 )
+                                                                           )
```
Comment (Thomas-Galtier) [inline mplc/multi_partner_learning.py:719]: Why use `np.max`?  It should  rather be : 
`partner_model.predict(np.expand_dims(x, axis=0))[y]`

Comment (Thomas-Galtier) [reply mplc/multi_partner_learning.py:719]: Also why `np.exand_dims` is needed?

File: mplc/multi_partner_learning.py [hunk 2]
```diff
@@ -654,3 +655,122 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
...
+                                norm += self.theta_[self.epoch_index, partner_index, y, z_i]
+                            self.theta[self.epoch_index + 1, partner_index, y, z] = temp / norm
+                        self.theta_[self.epoch_index + 1] = self.theta[self.epoch_index + 1]
+                        for idx, x in enumerate(x_batch):
+                            self.theta_[self.epoch_index + 1,
+                                        partner_index,
+                                        y,
+                                        np.argmax(y_batch[idx])] *= np.max(partner_model.predict(np.expand_dims(x,
+                                                                                                                axis=0)
+                                                                                                 )
```
Comment (Thomas-Galtier) [inline mplc/multi_partner_learning.py:741]: Same change here

File: mplc/contributivity.py [hunk 3]
```diff
@@ -938,6 +937,13 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
             end = timer()
             self.computation_time_sec = end - start
 
+    def Flip_label(self):
+        mpl = multi_partner_learning.MplLabelFlip(self.scenario)
+        mpl.fit()
+        self.contributivity_scores = [np.linalg.norm(
+            mpl.theta[-1][i] - np.identity(mpl.theta[-1][i].shape[0])
+        ) for i in range(len(self.scenario.partners_list))]
+
```
Comment (Thomas-Galtier) [inline mplc/contributivity.py:946]: As the contribution score is supposed to be high when there is no cheating I would rather put:
`  np.sqrt(mpl.theta[-1][i].shape[0]) - np.linalg.norm( mpl.theta[-1][i] - np.identity(mpl.theta[-1][i].shape[0]) )`

**Review by Thomas-Galtier — CHANGES_REQUESTED** [review]

File: mplc/multi_partner_learning.py (hunk 1 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning.py:719]: np.expand_dims is need because the dimension input of a conv_layer is 4. It is for batches. so, rather than have in entry x, I use a batch of size 1, [x]

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -163,6 +163,23 @@ Notebooks now have really efficient ways to share your code and display your out
 - [Binder](https://mybinder.org/) that lets you turn a Git repository into a collection of interactive notebooks!
 - [Voila](https://github.com/voila-dashboards/voila) is a newcomer that will allow to build dashboads for your presentations!
 
+#### 2.6 Release a new build 
+To release a new version on Pypi, go at the root of the repository, and trigger the build with pip.
+You will need all the dev-requirements installed.
+
+```bash
+
```
Comment (natct10) [inline CONTRIBUTING.md:171]:
```suggestion
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -163,6 +163,23 @@ Notebooks now have really efficient ways to share your code and display your out
...
+$pip install -r dev-requirements.txt
+
```
Comment (natct10) [inline CONTRIBUTING.md:173]:
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -66,6 +68,31 @@ Once a given scenario configured, it seemed useful to choose how the multi-partn
 
 Finally, with given scenarios and multi-partner learning approaches, we can address contributivity measurement approaches. See below sections [Configuration of contributivity measurement methods to be tested](./mplc/docs/documentation.md#scenario-parameters) and [Contributivity measurement approaches studied and implemented](./mplc/docs/documentation.md#configuration-of-contributivity-measurement-methods-to-be-tested).
 
+### Installation 
+
+##### Using pip
+
+```bash
+
```
Comment (natct10) [inline README.md:76]:
```suggestion
```

File: README.md [hunk 7]
```diff
@@ -66,6 +68,31 @@ Once a given scenario configured, it seemed useful to choose how the multi-partn
...
+$ pip install mplc
+
```
Comment (natct10) [inline README.md:78]:
```suggestion
```

File: README.md [hunk 8]
```diff
@@ -66,6 +68,31 @@ Once a given scenario configured, it seemed useful to choose how the multi-partn
...
+This installs the last packaged version on pypi. If your platform doesn't have a ready package, please open an issue to let us know.
+
+##### Build from Source 
+
+If you want to install mplc from the repository, make sure that you got the latest version of pip. 
+
+Then clone the repository, and trigger the installation using pip.
+
+```bash
+
```
Comment (natct10) [inline README.md:90]:
```suggestion
```

File: README.md [hunk 9]
```diff
@@ -66,6 +68,31 @@ Once a given scenario configured, it seemed useful to choose how the multi-partn
...
+$ git clone https://github.com/SubstraFoundation/distributed-learning-contributivity.git
+$ cd distributed-learning-contributivity
+$ pip install -e . 
+
```
Comment (natct10) [inline README.md:94]:
```suggestion
```

File: mplc/multi_partner_learning.py [hunk 10]
```diff
@@ -654,3 +655,151 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
...
+                    partner_model = partners_models[partner_index]
+                    x_batch = self.minibatched_x_train[partner_index][self.minibatch_index]
+                    y_batch = self.minibatched_y_train[partner_index][self.minibatch_index]
+
+                    predictions = partner_model.predict(x_batch)
+                    self.theta_ = predictions  # Initialize the theta_
+
+                    for idx, y in enumerate(y_batch):
+                        self.theta_[idx, :] *= self.theta[:, np.argmax(y)]
+                        self.theta_ = normalize(self.theta_, axis=1, norm='l1')
```
Comment (Thomas-Galtier) [inline mplc/multi_partner_learning.py:721]: Line 721 can be put out of the for loop

File: mplc/multi_partner_learning.py [hunk 11]
```diff
@@ -654,3 +655,151 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
...
+                    self.history_theta_[self.epoch_index][partner_index] = self.theta
```
Comment (Thomas-Galtier) [inline mplc/multi_partner_learning.py:722]:
```suggestion
                    self.history_theta_[self.epoch_index][partner_index] = self.theta_
```

File: mplc/multi_partner_learning.py [hunk 12]
```diff
@@ -654,3 +655,151 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
...
+
+                    self.theta = self.theta_.T.dot(y_batch)
+                    self.theta = normalize(self.theta, axis=1, norm='l1')
+
+                    self.history_theta[self.epoch_index][partner_index] = self.theta
+
+                    self.theta_ = predictions
+                    for idx, y in enumerate(y_batch):
+                        self.theta_[idx, :] *= self.theta[:, np.argmax(y)]
+                        self.theta_ = normalize(self.theta_, axis=1, norm='l1')
```
Comment (Thomas-Galtier) [inline mplc/multi_partner_learning.py:732]: Same thing, line 732 can be put out of the for loop
```suggestion
                    for idx, y in enumerate(y_batch):
                        self.theta_[idx, :] *= self.theta[:, np.argmax(y)]
                    self.theta_ = normalize(self.theta_, axis=1, norm='l1')
```

File: dev-requirements.txt [hunk 13]
```diff
@@ -10,4 +11,7 @@ loguru==0.4.1
 tensorflow==2.2.0
 ruamel.yaml==0.16.10
 pytest==5.4.3
-flake8==3.8.3
\ No newline at end of file
+flake8==3.8.3
+twine==3.2.0
+setuptools==50.3.0
+wheel==0.35.1
```
Comment (natct10) [inline dev-requirements.txt:17]: Isn't this related to another [PR](https://github.com/SubstraFoundation/distributed-learning-contributivity/pull/246/files)?

File: mplc/docs/documentation.md [hunk 14]
```diff
@@ -298,7 +298,76 @@ All methods available are:
...
 ```
 
-See below section [Contributivity measurement approaches studied and implemented](#contributivity-measurement-approaches-studied-and-implemented) for explanation of the different methods.  
+##### Configuration of contributivity measurement methods to be tested
+
+`methods`:  
+A declarative list `[]` of the contributivity measurement methods to be executed.
+All methods available are:
+
+```
```
Comment (natct10) [inline mplc/docs/documentation.md:307]:
```suggestion
```sh
```

File: setup.py [hunk 15]
```diff
@@ -8,7 +8,7 @@
 
 setup(
     name="mplc",
-    version='0.0.0.1',
+    version='0.1',
```
Comment (natct10) [inline setup.py:11]: Isn't this is part of another PR?

File: dev-requirements.txt (hunk 13 above)
Comment (arthurPignet) [reply dev-requirements.txt:17]: Yes it is 
I think I cherrypicked it but I cannot remember why

File: mplc/docs/documentation.md (hunk 14 above)
Comment (arthurPignet) [reply mplc/docs/documentation.md:307]: I will make this change in the other PR

File: mplc/contributivity.py [hunk 16]
```diff
@@ -938,6 +937,23 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
             end = timer()
             self.computation_time_sec = end - start
 
+    def Flip_label(self):
```
Comment (RomainGoussault) [inline mplc/contributivity.py:940]: first letter should be lower case for a method name: `flip_label`

File: mplc/docs/documentation.md [hunk 17]
```diff
@@ -298,7 +298,76 @@ All methods available are:
...
+- "Shapley values"
+- "Independent scores"
+- "TMCS"
+- "ITMCS"
+- "IS_lin_S"
+- "IS_reg_S"
+- "AIS_Kriging_S"
+- "SMCS"
+- "WR_SMC"
+```
```
Comment (RomainGoussault) [inline mplc/docs/documentation.md:317]: LFLIP Missing ?

File: mplc/multi_partner_learning.py [hunk 18]
```diff
@@ -654,3 +655,153 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
     )
 
     return mpl
+
+
+class MplLabelFlip(MultiPartnerLearning):
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning.py:660]: add link to paper / blog post

File: mplc/docs/documentation.md [hunk 19]
```diff
@@ -298,7 +298,76 @@ All methods available are:
...
+
+The methods are detailed below:
```
Comment (RomainGoussault) [inline mplc/docs/documentation.md:319]: LFLIP missing ?

**Review by RomainGoussault — CHANGES_REQUESTED** [review]

File: mplc/multi_partner_learning.py (hunk 18 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning.py:660]: There is no paper yet, nor blog post. Thomas and I are working on the paper.

File: mplc/multi_partner_learning.py [hunk 20]
```diff
@@ -654,3 +655,153 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
...
+            scenario.use_saved_weights,
+        )
+
+        self.epsilon = epsilon
+        self.K = scenario.dataset.num_classes
+        self.history_theta = [[None for _ in self.partners_list] for _ in range(self.epoch_count)]
+        self.history_theta_ = [[None for _ in self.partners_list] for _ in range(self.epoch_count)]
+        self.theta = [self.init_flip_proba() for _ in self.partners_list]
+        self.theta_ = [None for _ in self.partners_list]
+        # self.labels_map = LabelEncoder().fit_transform([str(y) for y in scenario.dataset.y_train])
```
Comment (natct10) [inline mplc/multi_partner_learning.py:683]: Is this going to be used some day? Just asking

File: mplc/contributivity.py [hunk 21]
```diff
@@ -62,18 +62,17 @@ def predict(self, x):
...
+        self.contributivity_scores = np.zeros(nb_partners)
+        self.scores_std = np.zeros(nb_partners)
+        self.normalized_scores = np.zeros(nb_partners)
         self.computation_time_sec = 0.0
         self.first_charac_fct_calls_count = 0
         self.charac_fct_values = {(): 0}
-        self.increments_values = []
-        for i in range(n):
-            self.increments_values.append(dict())
+        self.increments_values = [{} for _ in self.scenario.partners_list]
```
Comment (natct10) [inline mplc/contributivity.py:75]: Sure about this? Just asking

Comment (arthurPignet) [reply mplc/contributivity.py:75]: I just rewrite the initialization of self.increments_values. I don't know where it is use.

File: mplc/multi_partner_learning.py (hunk 20 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning.py:683]: Egh not really, it was an attempt to make it agnostic to label type. But this problem must be adressed with deeper changes in the code.

File: mplc/contributivity.py (hunk 21 above)
Comment (Thomas-Galtier) [reply mplc/contributivity.py:75]: It is used in the importance sampling estimation of the shapley values. If I remember well, I had a pointer issue when initializing the increment values as you did. I think it is better to keep it as was.

File: mplc/multi_partner_learning.py [hunk 22]
```diff
@@ -654,3 +655,152 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
...
+                    self.theta_[partner_index] = predictions  # Initialize the theta_
+
+                    for idx, y in enumerate(y_batch):
+                        self.theta_[partner_index][idx, :] *= self.theta[partner_index][:, np.argmax(y)]
+                    self.theta_[partner_index] = normalize(self.theta_[partner_index], axis=0, norm='l1')
+                    self.history_theta_[self.epoch_index][partner_index] = self.theta_[partner_index]
+
+                    self.theta[partner_index] = self.theta_[partner_index].T.dot(y_batch)
+                    self.theta[partner_index] = normalize(self.theta[partner_index], axis=1, norm='l1')
+                    logger.debug(f"(LFlip) theta : {self.theta[partner_index]}")
```
Comment (Thomas-Galtier) [inline mplc/multi_partner_learning.py:726]: If think we can remove this log, as we can access the theta with history

**Review by Thomas-Galtier — CHANGES_REQUESTED** [review]

File: mplc/multi_partner_learning.py (hunk 22 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning.py:726]: As you wish but it is a debug log, so it is not display by default

File: mplc/contributivity.py (hunk 21 above)
Comment (arthurPignet) [reply mplc/contributivity.py:75]: If you initialized this by doing [{}] * n, yes indeed you would have pointer issues. 
But with this change, it is exactly the same as before, without issue (tested)

**Review by Thomas-Galtier — APPROVED** [review]

File: mplc/docs/documentation.md [hunk 23]
```diff
@@ -367,6 +378,16 @@ The methods are detailed below:
     - `["Federated SBS linear"]` - Linear importance increase between computation rounds (1000th round weights 1000 times first round)
     - `["Federated SBS quadratic"]` - Quadratic importance increase between computation rounds (1000th round weights 10e6 times first round)
     - `["Federated SBS constant"]`- Constant importance increase between computation rounds (1000th round weights same as first round)
+    
+- [In progress] **Label Flipping**
```
Comment (bowni) [inline mplc/docs/documentation.md:382]: Juste remove "[In progress]"
```suggestion
- **Label Flipping**
```

File: mplc/docs/documentation.md [hunk 24]
```diff
@@ -367,6 +378,16 @@ The methods are detailed below:
...
+    
+    Label Flipping method provides a way to detect mislabelled datasets.
+    The main idea is, while training the model, to learn the probability of a label to be flipped in another, inside each partner dataset. 
+    Then, we flip the label of the noisy data to the most likely right label and train the main model on these likely right data.  
+    A contributivity measure can be inferred from partner's matrices of flip-probability, by computing the exponential inverse of the Frobenius distance to the identity. 
+    However this measure is to be handle carefully, the method is not designed specifically for contributivity measurement, but for mislabelled dataset detection. 
+    
+    - `["LFip]` - Label flipping method
```
Comment (bowni) [inline mplc/docs/documentation.md:390]:
```suggestion
    - `["LFip"]` - Label flipping method
```

**Review by bowni — APPROVED** [review]

**Review by RomainGoussault — APPROVED** [review]
👍
